src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,344 @@
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given IP
+    if (rate_limit_entities.count(rate_limit_resource_t{entity_type,entity}) == 0) {
+        return false;
+    }
+
+    // Remove the rule from the rule store
+    rule_store.erase(rate_limit_entities.at(rate_limit_resource_t{entity_type,entity})->id);
+
+    // Remove the rule from the entity rate limits map
+    rate_limit_entities.erase(rate_limit_resource_t{entity_type,entity});
+
+    return true;
+
+}
+
+void RateLimitManager::temp_ban_entity(const rate_limit_resource_t& entity, const int64_t number_of_days) {
+    // lock mutex
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+    temp_ban_entity_wrapped(entity, number_of_days);
+}
+
+bool RateLimitManager::ban_entities_permanently(const RateLimitedEntityType entity_type, const std::vector<std::string> &entities) {
+    // lock mutex
+    std::unique_lock < std::shared_mutex > lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given IP addresses
+    for(const auto& entity : entities) {
+        if (rate_limit_entities.count(rate_limit_resource_t{entity_type,entity}) > 0) {
+            return false;
+        }
+    }
+
+    // add rate limit for entity
+    rule_store.insert({last_rule_id,rate_limit_rule_t{last_rule_id,RateLimitAction::BLOCK, entity_type,entities, rate_limit_max_requests_t{0, 0}, -1, -1}});
+
+    // Add rule from rule store to rate limit rule pointer
+    for(const auto &entity : entities) {
+        rate_limit_entities.insert({rate_limit_resource_t{entity_type,entity}, &rule_store.at(last_rule_id)});
+    }
+
+    last_rule_id++;
+    return true;
+}
+
+
+
+bool RateLimitManager::is_rate_limited(const std::vector<rate_limit_resource_t> &entities) {
+    // lock mutex
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+    
+    // Check if any of the entities has rule
+    for(const auto &entity : entities) {
+
+        if (rate_limit_entities.count(entity) == 0) {
+            continue;
+        }
+
+        const auto& rule = *(rate_limit_entities.at(entity));
+
+        if (rule.action == RateLimitAction::BLOCK) {
+            return true;
+        }
+        else if(rule.action == RateLimitAction::ALLOW) {
+            return false;
+        }
+        else if(throttled_entities.count(entity) > 0 && throttled_entities.at(entity).is_banned) {
+            
+            // Check if ban duration is not over
+            if (throttled_entities.at(entity).throttling_to > std::chrono::system_clock::to_time_t(std::chrono::system_clock::now())) {
+                
+                // Delete request counts for the entity
+                if(rate_limit_request_counts.contains(entity)) {
+                    rate_limit_request_counts.erase(entity);
+                }
+
+                return true;
+            }
+ 
+            // Remove ban
+            throttled_entities.at(entity).is_banned = false;
```

Comment:
Why should we set a flag here instead of removing entity from the `throttled_entities` map? In fact, I think the `is_banned` flag is redundant here because a presence of an entity in this map already means it's throttled.